src/uct/ib/dc/dc_mlx5_ep.c

```diff
@@ -1061,13 +1065,18 @@ uct_dc_mlx5_iface_dci_do_pending_tx(ucs_arbiter_t *arbiter,
     if (status == UCS_INPROGRESS) {
         return UCS_ARBITER_CB_RESULT_NEXT_GROUP;
     }
-    if (!uct_dc_mlx5_iface_dci_ep_can_send(ep) ||
-        uct_dc_mlx5_ep_fc_wait_for_grant(ep)) {
-        /* Deschedule the group even if FC is the only resource, which
-         * is missing. It will be scheduled again when credits arrive. */
 
-        /* TODO: only good for dcs policy */
-        return UCS_ARBITER_CB_RESULT_DESCHED_GROUP;
+    if (!uct_dc_mlx5_iface_dci_ep_can_send(ep)) {
+        /* Deschedule the group even if FC is the only resource, which
+         * is missing. It will be scheduled again when credits arrive.
+         * We can't desched group with rand policy if non FC resources are
+         * missing, since it's never scheduled again. */
+        if (uct_dc_mlx5_iface_is_dci_rand(iface) &&
+            uct_rc_fc_has_resources(&iface->super.super, &ep->fc)) {
+            return UCS_ARBITER_CB_RESULT_RESCHED_GROUP;
+        } else {
+            return UCS_ARBITER_CB_RESULT_DESCHED_GROUP;
```

Comment:
the condition here is changed, even for DCS policy. What is the reason? Maybe should add assertions on  uct_dc_mlx5_ep_fc_wait_for_grant()?

Reply:
not sure I understand the question. For dcs* policies we desched the group always (like it was before).
For rand policy we:
- desched the group of FC resources are missing, because the group will be scheduled again (for any policy) when FC grant arrives
- resched the group if non FC resources are missing, because it will not be scheduled again upon TX completions

Reply:
is it possible that uct_dc_mlx5_iface_dci_ep_can_send(ep) is false, and uct_dc_mlx5_ep_fc_wait_for_grant(ep) is true?
in this case, before the change, it returned DESCHED, and after, it would return STOP (for dcs*)

Reply:
i saw you added it in 70df3ea

Reply:
not sure why we checked for `uct_dc_mlx5_ep_fc_wait_for_grant` originally. It does not meant that FC resources are exhausted, it means that grant request was sent. But `uct_dc_mlx5_iface_dci_ep_can_send(ep)` checks for all ep related resources: QP and FC. Thus:
- if ep can send - it means we don't have iface resources, we have to stop
- if ep can't send we need to deschedule the group (or reschedule in case of FC)

Reply:
i think the idea was that if EP is waiting for grant, it couldn't be flushed, to avoid segfault when getting grant message

Reply:
but we have list of destroyed ep now. 
I guess this particular change was a mistake in 70df3ea. I should had left check for fc window instead (replaced with inline func which is not exactly the same)

Reply:
ok